docs/ibc/middleware/develop.md

```diff
@@ -0,0 +1,230 @@
+
+Middleware allows developers to define the extensions as seperate modules that can wrap over the base application. This middleware can thus perform its own custom logic, and pass data into the application so that it may run its logic without being aware of the middleware's existence. This allows both the application and the middleware to implement its own isolated logic while still being able to run as part of a single packet flow.
+
+## Pre-requisite Readings
+
+- [IBC Overview](../overview.md) {prereq}
+- [IBC default integration](../integration.md) {prereq}
+- [IBC Application Developer Guide](../apps.md) {prereq}
+
+## Definitions
+
+`Middleware`: A self-contained module that sits between core IBC and an underlying IBC application during packet execution. All messages between core IBC and underlying application must flow through middleware, which may perform its own custom logic.
+
+`Underlying Application`: An underlying application is the application that is directly connected to the middleware in question. This underlying application may itself be middleware that is chained to a base application.
+
+`Base Application`: A base application is an IBC application that does not contain any middleware. It may be nested by 0 or multiple middleware to form an application stack.
+
+`Application Stack (or stack)`: A stack is the complete set of application logic (middleware(s) +  base application) that gets connected to core IBC. A stack may be just a base application, or it may be a series of middlewares that nest a base application.
+
+## Create a custom IBC Middleware
+
+IBC Middleware will wrap over an underlying IBC application and sits between core IBC and the application. It has complete control in modifying any message coming from IBC to the application, and any message coming from the application to core IBC. Thus, middleware must be completely trusted by chain developers who wish to integrate them, however this gives them complete flexibility in modifying the application(s) they wrap.
+
+#### Interfaces
+
+```typescript
+// Middleware implements the ICS26 Module interface
+interface Middleware extends ICS26Module {
+    app: ICS26Module // middleware has acccess to an underlying application which may be wrapped by more middleware
+    ics4Wrapper: ICS4Wrapper // middleware has access to ICS4Wrapper which may be core IBC Channel Handler or a higher-level middleware that wraps this middleware.
+}
+```
+
+```typescript
+// This is implemented by ICS4 and all middleware that are wrapping base application.
+// The base application will call `sendPacket` or `writeAcknowledgement` of the middleware directly above them
+// which will call the next middleware until it reaches the core IBC handler.
+interface ICS4Wrapper {
+    sendPacket(packet: Packet)
+    writeAcknowledgement(packet: Packet, ack: Acknowledgement)
+}
+```
+
+### Implement `IBCModule` interface and callbacks
```

Comment:
This should define the `IBCModule` type as well (indicating it should be separate from AppModule`). I would even recommend creating an `ibc_module.go` file